File: aoc-2021/packages/aoc-rust/src/day_02.rs
use std::fs::File;
use std::io::{prelude::*, BufReader};

fn part_one() -> i32 {
    let data = BufReader::new(File::open("../../input/day_02.in").unwrap())
        .lines()
        .filter_map(|l| l.ok())
        .fold((0_i32, 0_i32), |x_y, line| {
            let data: Vec<&str> = line.split(' ').collect();
            let val = data[1].parse::<i32>().unwrap();
            match data[0] {
                "forward" => (x_y.0 + val, x_y.1),
                "up" => (x_y.0, x_y.1 - val),
                "down" => (x_y.0, x_y.1 + val),
                _ => x_y,
            }
        });
    data.0 * data.1
}

fn part_two() -> i32 {
    let data = BufReader::new(File::open("../../input/day_02.in").unwrap())
        .lines()
        .filter_map(|l| l.ok())
        .fold((0_i32, 0_i32, 0_i32), |x_y_aim, line| {
            let data: Vec<&str> = line.split(' ').collect();
            let val = data[1].parse::<i32>().unwrap();
            match data[0] {
                "forward" => (x_y_aim.0 + val, x_y_aim.1 + x_y_aim.2 * val, x_y_aim.2),
                "up" => (x_y_aim.0, x_y_aim.1, x_y_aim.2 - val),
                "down" => (x_y_aim.0, x_y_aim.1, x_y_aim.2 + val),
                _ => x_y_aim,
            }
        });
    data.0 * data.1
}
pub fn show() {
    println!("- Answer part one: {}", part_one());
    println!("- Answer part two: {}", part_two());
}

File: aoc-go/2021/day_02.go
package solutions

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func d02_partOne() int {
	file, _ := os.Open("../../input/day_02.in")
	defer file.Close()

	scanner := bufio.NewScanner(file)
	x := 0
	y := 0
	for scanner.Scan() {
		data := strings.Split(scanner.Text(), " ")
		value, _ := strconv.Atoi(data[1])
		switch data[0] {
		case "forward":
			x += value
		case "up":
			y -= value
		case "down":
			y += value
		}
	}
	return x * y
}

func d02_partTwo() int {
	file, _ := os.Open("../../input/day_02.in")
	defer file.Close()

	scanner := bufio.NewScanner(file)
	x := 0
	y := 0
	aim := 0
	for scanner.Scan() {
		data := strings.Split(scanner.Text(), " ")
		value, _ := strconv.Atoi(data[1])
		switch data[0] {
		case "forward":
			x += value
			y += aim * value
		case "up":
			aim -= value
		case "down":
			aim += value
		}
	}
	return x * y
}

func ShowD02() {
	fmt.Printf("- Answer part one: %d\n", d02_partOne())
	fmt.Printf("- Answer part two: %d\n", d02_partTwo())
}
